### PR TITLE
Initial Listing state machine setup

### PR DESCRIPTION
**Pull request by software-project**
# Pull Request Template

## Issue

Addresses #1582

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Adds state machine with guards and validation.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [ ] This change is a refactor/address technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Please describe the tests that you ran to verify your changes. Provide instructions so we can review. Please also list any relevant details for your test configuration

- [ ] Desktop View
- [ ] Mobile View
- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes

### CODE REVIEW

**Comment by software-project** [comment]
@seanmalbert @pbn4 @emilyjablonski @dominikx96 @jaredcwhite  This is an initial state machine for listings. 
What's working:
- state transitions with guards
- it passes an error when it's one (we need to figure out the best way to handle them and pass them to frontend) @pbn4 do you know if there is a standard way to handle it?

I still need to add some tests and schemas for the validation are not complete. Also there is an caching issue, so even when state is changed, view remains unchanged. @seanmalbert do you know what it might be? Let me know what do you think and I'll make updates this week.

File: backend/core/src/listings/types/listing-status-enum.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export enum ListingStatus {
   active = "active",
-  pending = "pending",
   closed = "closed",
+  new = "new",
```
Comment (emilyjablonski) [inline backend/core/src/listings/types/listing-status-enum.ts:4]: What does `new` represent as a status?

**Comment by seanmalbert** [comment]
@dominikx96 ,
Can you please review and make sure this makes sense to you. You'll be taking the end result of this to flesh out the rest of the state validation.

File: backend/core/src/listings/types/listing-status-enum.ts (hunk 1 above)
Comment (software-project) [reply backend/core/src/listings/types/listing-status-enum.ts:4]: It's an initial state for the state machine. Pending state requires already some validations to pass, so we need an initial point.

**Review by pbn4 — CHANGES_REQUESTED** [review]
I like the idea of using xstate for state management and condition guards, it will handling existing and future transition in listing state scalable and easy to maintain. 

Regarding using `yup` I think it represent a different convention than what we already have with `class-validator` and we should stick with the latter to have consistent error messages. `ValidationPipe` can be injected through constructor to `Listings.service.ts` and it exposes `transform` method specifically for inline validation. I think the solution should a mix of `xstate` with transition guards and `ValidationPipe.transform` inline depending (triggered in transition guards) on what Listing state we would like to transition to. It implies that we need separate classes like ListingPublished defined with all the `class-validator` decorators in place with appropriate logic for a given state. For more complex validations (e.g. including DB access) it's possible to do DI in custom validation decorator classes like presented [here](https://github.com/nestjs/nest/issues/528). 

@seanmalbert I can take over this PR.

**Comment by jaredcwhite** [comment]
@pbn4 I thought the point of using Yup is we can validate the JSON payload entirely on the client. Doing it again on the backend is just extra hygiene at that point. Using validation concerns in Nest wouldn't allow us to do client-side validation easily if I'm not mistaken.

**Comment by software-project** [comment]
@dominikx96 @jaredcwhite yes, I agree with you, that returning an error object, so we can assign errors to field would be my prefered way, but I could find a way to do so with yup. One thing to check also, cause I don't remember if I fixed it, was an issue where yup was returning only the first validation error. As there I quite a few things to resolve here I'm happy if @pbn4 could take it over.

**Comment by seanmalbert** [comment]
This work should now be covered with #1796

**Pull request closed without merging**